sdk/communication/communication-common/test/public/node/entraCommunicationTokenCredential.spec.ts

```diff
@@ -58,6 +63,136 @@ describe("Entra CommunicationTokenCredential", function () {
+
+    const tokenResult = (await credential.getToken()).token;
+    assert.strictEqual(tokenResult, acsToken);
+    assert.isTrue(getTokenSpy.alwaysCalledWithExactly(defaultScopes, undefined));
+    assert.isTrue(getTokenSpy.callCount > 0);
+    assert.isTrue(scope.isDone());
+    getTokenSpy.restore();
+  });
+
+  it("Token exchange is not called again when Entra token stays the same", async function () {
+    let scope = successApiMock();
+    const entraTokenCredentialOptions: EntraCommunicationTokenCredentialOptions = {
+      resourceEndpoint: resourceEndpoint,
+      tokenCredential: tokenCredential,
+      scopes,
+    };
+    const entraTokenCredential = new EntraTokenCredential(entraTokenCredentialOptions);
+    const exchangeTokenSpy = sinon.spy(entraTokenCredential as any, "exchangeEntraToken");
+
+    let tokenResult = (await entraTokenCredential.getToken()).token;
+    assert.strictEqual(tokenResult, acsToken);
+    assert.isTrue(exchangeTokenSpy.callCount > 0);
+    assert.isTrue(scope.isDone());
+
+    scope = successApiMock();
+    exchangeTokenSpy.restore();
+    exchangeTokenSpy.resetHistory();
+
+    tokenResult = (await entraTokenCredential.getToken()).token;
+    assert.strictEqual(tokenResult, acsToken);
+    assert.isTrue(exchangeTokenSpy.callCount === 0);
+    exchangeTokenSpy.restore();
+  });
+
+  it("Token exchange gets called again when Entra token changes", async function () {
+    let scope = successApiMock();
+    let entraTokenCredentialOptions: EntraCommunicationTokenCredentialOptions = {
+      resourceEndpoint: resourceEndpoint,
+      tokenCredential: tokenCredential,
+      scopes,
+    };
+    let entraTokenCredential = new EntraTokenCredential(entraTokenCredentialOptions);
+    let exchangeTokenSpy = sinon.spy(entraTokenCredential as any, "exchangeEntraToken");
+
+    let tokenResult = (await entraTokenCredential.getToken()).token;
+    assert.strictEqual(tokenResult, acsToken);
+    assert.isTrue(exchangeTokenSpy.callCount > 0);
+    assert.isTrue(scope.isDone());
+
+    const apiMockNewToken = () =>
+      nock(resourceEndpoint)
+        .post(tokenExchangePath)
+        .matchHeader("Authorization", `Bearer ${newEntraToken}`);
+
+    const successApiMockNewToken = () =>
+      apiMockNewToken().reply(200, {
+        accessToken: {
+          token: acsToken,
+          expiresOn: new Date(Date.now() + 60 * 60 * 1000).toISOString(),
+        },
+      });
+
+    scope = successApiMockNewToken();
+    const acsTokenCredential: TokenCredential = {
+      getToken: async (_scopes: string, _options?: GetTokenOptions) => {
+        return {
+          token: newEntraToken,
+          expiresOnTimestamp: Date.now() + 60 * 60 * 1000,
+        };
+      },
+    };
+    entraTokenCredentialOptions = {
+      resourceEndpoint: resourceEndpoint,
+      tokenCredential: acsTokenCredential,
+      scopes,
+    };
+    entraTokenCredential = new EntraTokenCredential(entraTokenCredentialOptions);
```

Comment:
You're now using a completely different EntraTokenCredential. I would write this test case a bit differently: you can implement your `TokenCredential` in a way that you return `token: passNewToken ? newEntraToken : entraToken`. And then you call the outer entraTokenCredential.getToken() once, validate that it exchanged, call it again, validate that it didn't exchange, and then set passNewToken to true, call getToken again and validate that it then exchanged again.